library/core/src/str/count.rs

```diff
@@ -0,0 +1,116 @@
+//! Code for efficiently counting the number of `char`s in a UTF-8 encoded
+//! string.
+//!
+//! Broadly, UTF-8 encodes `char`s as a "leading" byte which begins the `char`,
+//! followed by some number (possibly 0) of continuation bytes.
+//!
+//! The leading byte can have a number of bit-patterns (with the specific
+//! pattern indicating how many continuation bytes follow), but the continuation
+//! bytes are always in the format `0b10XX_XXXX` (where the `X`s can take any
+//! value). That is, the most significant bit is set, and the second most
+//! significant bit is unset.
+//!
+//! To count the number of characters, we can just count the number of bytes in
+//! the string which are not continuation bytes, which can be done many bytes at
+//! a time fairly easily.
+//!
+//! Note: Because the term "leading byte" can sometimes be ambiguous (for
+//! example, it could also refer to the first byte of a slice), we'll often use
+//! the term "non-continuation byte" to refer to these bytes in the code.
+
+pub(super) fn count_chars(s: &str) -> usize {
+    // For correctness, `CHUNK_SIZE` must be:
+    // - Less than or equal to 255, otherwise we'll overflow bytes in `counts`.
+    // - A multiple of `UNROLL_INNER`, otherwise our `break` inside the
+    //   `body.chunks(CHUNK_SIZE)` loop.
+    //
+    // For performance, `CHUNK_SIZE` should be:
+    // - Relatively cheap to `%` against.
```

Comment:
> That said, we don't `% CHUNK_SIZE` anywhere, do we?

`chunks` does in some places: https://doc.rust-lang.org/src/core/slice/iter.rs.html#1400
This could probably be added.
```suggestion
    // - Relatively cheap to `%` against for `chunks`.
```

Reply:
>  I'd also try 256 because that might be even faster.

Using chunk sizes >255 would allow overflowing the character counters packed into the usizes AFAIU, so this can't be done.

> chunks does in some places

Right, but nothing here ends up calling `size_hint`, neither directly or indirectly; and definitely does not do so often enough to warrant a warning like this. Constructing `array_chunk`s does a division though, and it has similar properties to the modulo operation, so perhaps this meant a `/` instead of `%`?

Reply:
Yeah, this should have indicated a division. I'll fix it when I update the code (sorry for the delay here)